jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java

```diff
@@ -548,27 +550,28 @@ protected void handleException(Throwable x)
             {
                 // Handle error normally
                 _request.setHandled(true);
-                _request.setAttribute(RequestDispatcher.ERROR_EXCEPTION, x);
-                _request.setAttribute(RequestDispatcher.ERROR_EXCEPTION_TYPE, x.getClass());
+                _request.setAttribute(RequestDispatcher.ERROR_EXCEPTION, failure);
+                _request.setAttribute(RequestDispatcher.ERROR_EXCEPTION_TYPE, failure.getClass());
 
                 if (isCommitted())
                 {
-                    abort(x);
+                    abort(failure);
                     if (LOG.isDebugEnabled())
-                        LOG.debug("Could not send response error 500, already committed", x);
+                        LOG.debug("Could not send response error 500, already committed", failure);
                 }
                 else
                 {
                     _response.setHeader(HttpHeader.CONNECTION.asString(), HttpHeaderValue.CLOSE.asString());
 
-                    if (x instanceof BadMessageException)
+                    Throwable cause = unwrap(failure);
+                    if (cause instanceof BadMessageException)
                     {
-                        BadMessageException bme = (BadMessageException)x;
+                        BadMessageException bme = (BadMessageException)cause;
                         _response.sendError(bme.getCode(), bme.getReason());
                     }
-                    else if (x instanceof UnavailableException)
+                    else if (cause instanceof UnavailableException)
                     {
-                        if (((UnavailableException)x).isPermanent())
+                        if (((UnavailableException)cause).isPermanent())
                             _response.sendError(HttpStatus.NOT_FOUND_404);
```

Comment:
It is really correct to send back `404` if `isPermanent()`? `4xx` are client errors and `UnavailableException` is not - perhaps a `500` is better?

Reply:
We send 404 because a permanent failure of a handler is the same as no handler at all.  It may be unavailable for perfectly normal reasons and thus a 500 is not correct.   Anyway, it's been like this for a while so I'm going to leave it as is.